File: Programmers/Level2/124_나라의_숫자/main.py
def solution(n):
    answer = ''
    stack = []

    while n > 0:
        # n을 3으로 나눈 나머지를 stack에 넣음
        stack.append(n%3)
        # n이 3의 배수이면 n // 3 에 1을 더 빼서 3의 배수가 아니게 만듦
        if n % 3 == 0:
            n = n // 3 - 1
            continue
        n //= 3
        
    # stack의 원소를 하나씩 빼면서 string으로 변환
    while stack:
        s = stack.pop()
        # n이 3의 배수일 경우 가장 끝자리는 무조건 "4"
        if s == 0:
            answer += "4"
        else:
            answer += str(s)
            
    return answer

print(solution(1), "1")
print(solution(2), "2")
print(solution(3), "4")
print(solution(4), "11")
print(solution(5), "12")
print(solution(6), "14")
print(solution(7), "21")
print(solution(8), "22")
print(solution(9), "24")
print(solution(10), "41")

File: Programmers/Level2/124_나라의_숫자/main.java
/*
- n을 3으로 나눈 나머지가 0일 땐 4, 1일 땐 1, 2일 땐 2에 대조된다.
- 수를 answer에 더해준 후, n-1을 3으로 나눈 값이 다음 n이 된다.
- 1의 자리 수부터 answer에 더해준 후 마지막에 뒤집는다.
*/

import java.lang.StringBuffer;

class Solution {
    public String solution(int n) {
        String answer = "";
        
        String[] num = {"4","1","2"};
        while (n > 0) {
            int m = n%3;
            answer += num[m];
            n = (n-1)/3;
        }
        
        StringBuffer sb = new StringBuffer(answer);
        String reverse_ans = sb.reverse().toString();
        return reverse_ans;
    }
}